### PR TITLE
kore-rpc server

### LINKED ISSUE #3066: Kore-server: setting up the JSON-RPC server

### PR DESCRIPTION
**Pull request by goodlyrottenapple**
Fixes #3066

This is a skeleton of a JSON RPC server API, outlined in #3061

### CODE REVIEW

File: kore/src/Kore/JsonRpc.hs [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        }
+    | HaltStuck
+        { depth :: !Int
+        }
+    | HaltDepthBound
+    | HaltPatternMatch
+        { depth :: !Int
+        , matches :: ![PatternMatch]
+        }
+    deriving stock (Show, Eq)
```
Comment (ana-pantilie) [inline kore/src/Kore/JsonRpc.hs:109]: Won't this lead to partial record accessors? I'd avoid that if possible.

File: kore/src/Kore/JsonRpc.hs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    Step StepRequest{} -> pure $ Right $ Step $ StepResult []
+    Implies _ -> pure $ Right $ Implies undefined
+    Simplify _ -> pure $ Right $ Simplify undefined
+    -- this case is only reachable if the cancel appeared as part of a batch request
+    Cancel -> pure $ Left $ ErrorObj "Cancel request unsupported in batch mode" (-32001) Null
+
+runServer :: IO ()
+runServer = do
+    runStderrLoggingT $ do
+        let ss = serverSettings 31337 "*"
```
Comment (ana-pantilie) [inline kore/src/Kore/JsonRpc.hs:213]: What does this number mean?

File: kore/src/Kore/JsonRpc.hs [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        jsonrpcTCPServer V2 False ss srv
+
+srv :: MonadLoggerIO m => JSONRPCT m ()
+srv = do
+    reqQueue <- liftIO $ atomically newTChan
+    qs <- ask
+    let sendResponses r = runStderrLoggingT $ runReaderT (sendBatchResponse r) qs
```
Comment (ana-pantilie) [inline kore/src/Kore/JsonRpc.hs:220]: Can some of the `let` binding be moved to the `where` clause? I think it's nicer to see the main logic in the body of the top level function and abstract away implementation details to the `where` clause.

**Review by ana-pantilie — COMMENTED** [review]
I just have a few comments/questions.

File: kore/src/Kore/JsonRpc.hs (hunk 2 above)
Comment (goodlyrottenapple) [reply kore/src/Kore/JsonRpc.hs:213]: this is just a random port, i will make it a parameter to `runServer`

File: kore/src/Kore/JsonRpc.hs [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        Simplify payload -> toJSON payload
+
+respond :: MonadIO m => Respond (API 'Req) m (API 'Res)
+respond = \case
+    Execute _ -> pure $ Right $ Execute undefined
+    Step StepRequest{} -> pure $ Right $ Step $ StepResult []
+    Implies _ -> pure $ Right $ Implies undefined
+    Simplify _ -> pure $ Right $ Simplify undefined
+    -- this case is only reachable if the cancel appeared as part of a batch request
+    Cancel -> pure $ Left $ ErrorObj "Cancel request unsupported in batch mode" (-32001) Null
```
[the remark this answers is not in the document]
Comment (goodlyrottenapple) [reply kore/src/Kore/JsonRpc.hs:208]: As per the spec https://www.jsonrpc.org/specification

-32000 to -32099 | Server error | Reserved for implementation-defined server-errors.
-- | -- | --

File: kore/src/Kore/JsonRpc.hs (hunk 1 above)
Comment (goodlyrottenapple) [reply kore/src/Kore/JsonRpc.hs:109]: Yes. What would you suggest instead? I wanted to keep close to the discussed spec, i.e.

```
{ "reason": "branching", "depth": DEPTH }
{ "reason": "stuck", "depth": DEPTH }
{ "reason": "depth-bound" }
{ "reason": "pattern-match", "depth": DEPTH, "matches": [ { "pattern": INT_INDEX_OF_PATTERN_THAT_MATCHED, "substitution": MATCHING_SUBSTITUTION } ] }
```

**Pull request merged**